### PR TITLE
[JAX] Enhance Dropout in TransformerLayer.

### PR DESCRIPTION
**Pull request by mingxu1067**
1. Fixed missing setup of dropout RNG key in TransformerLayer and LayerNormMLP.
2. Allowing seperated dropout rate for FC1's output and other hiddens.
3. Fix the wrong fp8 sclae issue in `_update_fp8_metas_impl`.

### CODE REVIEW

File: transformer_engine/jax/flax/transformer.py [hunk 1]
```diff
@@ -1078,6 +1084,8 @@ def __call__(self,
         else:
             mha_name = 'self_attention'
 
+        inputs = _with_sharding_constraint(inputs, (BATCH_AXES, SEQLEN_AXES, HIDDEN_AXES))
```
Comment (nouiz) [inline transformer_engine/jax/flax/transformer.py:1087]: Do you want to include those WSC in this PR or this belong to another PR?

File: transformer_engine/jax/flax/transformer.py [hunk 2]
```diff
@@ -1156,6 +1165,8 @@ def hidden_dropout(x, deterministic):
             y = hidden_dropout(y, deterministic)
             mlp_input = y + residual
 
+        mlp_input = _with_sharding_constraint(mlp_input, (BATCH_AXES, SEQLEN_AXES, HIDDEN_AXES))
```
Comment (nouiz) [inline transformer_engine/jax/flax/transformer.py:1168]: idem

File: tests/jax/test_praxis_layers.py [hunk 3]
```diff
@@ -1007,6 +1009,7 @@ def generate_praxis_p_and_flax_cls(self, dtype, attrs):
                            layernorm_type=layernorm_type,
                            hidden_dropout=hidden_dropout,
                            attention_dropout=attention_dropout,
+                           intermediate_dropout=intermediate_dropout,
```
Comment (nouiz) [inline tests/jax/test_praxis_layers.py:1012]: Do you test different value for this new parameter?

File: transformer_engine/jax/flax/transformer.py (hunk 1 above)
Comment (mingxu1067) [reply transformer_engine/jax/flax/transformer.py:1087]: Yes, adding WSC has been in TODO for a while, not regarding to CP.

File: tests/jax/test_praxis_layers.py (hunk 3 above)
Comment (mingxu1067) [reply tests/jax/test_praxis_layers.py:1012]: Not in UT, but will in Pax with 5B models

**Comment by mingxu1067** [comment]
@nouiz @ksivaman @timmoon10,
This PR also fixed a cirtial issue regarding update_fp8_meta.

**Comment by mingxu1067** [comment]
Previous failed tests are passed:
```
================================================================================= test session starts =================================================================================
platform linux -- Python 3.10.12, pytest-7.4.2, pluggy-1.3.0
rootdir: /mingh_wk/TransformerEngine
plugins: typeguard-4.1.3, jaxtyping-0.2.21, anyio-4.0.0
collected 6 items                                                                                                                                                                     

test_helper.py ......                                                                                                                                                           [100%]

================================================================================== warnings summary ===================================================================================
../../../../usr/local/lib/python3.10/dist-packages/tensorflow/__init__.py:29
  /usr/local/lib/python3.10/dist-packages/tensorflow/__init__.py:29: DeprecationWarning: The distutils package is deprecated and slated for removal in Python 3.12. Use setuptools or check PEP 632 for potential alternatives
    import distutils as _distutils

../../../../usr/local/lib/python3.10/dist-packages/tensorflow/python/framework/dtypes.py:205
  /usr/local/lib/python3.10/dist-packages/tensorflow/python/framework/dtypes.py:205: DeprecationWarning: `np.bool8` is a deprecated alias for `np.bool_`.  (Deprecated NumPy 1.24)
    np.bool8: (False, True),

../../../../usr/local/lib/python3.10/dist-packages/flatbuffers/compat.py:19
  /usr/local/lib/python3.10/dist-packages/flatbuffers/compat.py:19: DeprecationWarning: the imp module is deprecated in favour of importlib and slated for removal in Python 3.12; see the module's documentation for alternative uses
    import imp

../../../../usr/local/lib/python3.10/dist-packages/tensorboard/compat/tensorflow_stub/dtypes.py:326
  /usr/local/lib/python3.10/dist-packages/tensorboard/compat/tensorflow_stub/dtypes.py:326: DeprecationWarning: `np.bool8` is a deprecated alias for `np.bool_`.  (Deprecated NumPy 1.24)
    np.bool8: (False, True),

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
============================================================================ 6 passed, 4 warnings in 9.00s ============================================================================
root@preos0009:/mingh_wk/TransformerEngine/tests/jax# pytest test_layer.py 
================================================================================= test session starts =================================================================================
platform linux -- Python 3.10.12, pytest-7.4.2, pluggy-1.3.0
rootdir: /mingh_wk/TransformerEngine
plugins: typeguard-4.1.3, jaxtyping-0.2.21, anyio-4.0.0
collected 528 items                                                                                                                                                                   

test_layer.py ................................................................................................................................................................. [ 30%]
............................................................................................................................................................................... [ 63%]
............................................................................................................................................................................... [ 96%]
.................                                                                                                                                                               [100%]

================================================================================== warnings summary ===================================================================================

```

File: transformer_engine/jax/flax/module.py [hunk 4]
```diff
@@ -739,6 +739,9 @@ class LayerNormMLP(TransformerEngineBase):
     activations: Sequence[Union[str, Callable]], default = ('relu',)
         The sequence of activation functions to apply after the first linear transformation.
         Each activation has its own transformation layer.
+    intermediate_dropout_rng_name: str, default = 'dropout'
+        The key in given RNGs via flax.linen.Module.apply that for
+        generating Dropout masks in the Multi-Head Attention.
```
Comment (zlsh80826) [inline transformer_engine/jax/flax/module.py:744]: Is `intermediate_dropout_rng_name` for `Multi-Head Attention`?

**Review by zlsh80826 — APPROVED** [review]

File: transformer_engine/jax/flax/module.py (hunk 4 above)
Comment (mingxu1067) [reply transformer_engine/jax/flax/module.py]: It is typo, fixed.

**Comment by mingxu1067** [comment]
@timmoon10 @ksivaman , testing on H100 is timeout, should we relaunch a CI pipeline, or I can test on local side then share the log here?
